### PR TITLE
Add renderMode to boolean question types [DDP-7709]

### PR DESCRIPTION
**Pull request by bskinner**
DDP-7709

## Context

This PR adds the BooleanRenderMode type, it's associated in-database object, and updated to the API. This continues work started in PR #1345. If the renderMode is unspecified, all existing studies should default to the `RADIO_BUTTONS` render mode.

There should be no visible changes in behavior for any studies currently in production, and the new `CHECKBOX` render mode should only be used when explicitly set in the study configuration.

## How do we demo these changes?

The changes include a set of test questions in a new Basil activity.

### CODE REVIEW

File: pepper-apis/dss-database/src/main/resources/db-changes/schema/DDP-7709-add-boolean-render-mode.xml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                   http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.5.xsd">
+
+    <changeSet author="bskinner" id="add-boolean-render-mode">
+        <addColumn tableName="boolean_question">
+            <column name="boolean_render_mode_id" type="bigint" valueComputed="(select boolean_render_mode_id from boolean_render_mode where boolean_render_mode_code = 'RADIO_BUTTONS')">
+                <constraints nullable="false" references="boolean_render_mode(boolean_render_mode_id)" foreignKeyName="boolean_question_boolean_render_mode_fk"/>
```
Comment (MocanaAtBroad) [inline pepper-apis/dss-database/src/main/resources/db-changes/schema/DDP-7709-add-boolean-render-mode.xml:10]: Nice! I would have just taken null to mean RADIO_BUTTON, but this is better!

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dao/JdbiBooleanQuestion.java [hunk 2]
```diff
@@ -1,14 +1,21 @@
...
 
 public interface JdbiBooleanQuestion extends SqlObject {
 
-    @SqlUpdate("insert into boolean_question (question_id,true_template_id,false_template_id)"
-            + " values(:questionId,:trueTemplateId,:falseTemplateId)")
+    @UseStringTemplateSqlLocator
+    @SqlUpdate("insertByRenderModeId")
     int insert(@Bind("questionId") long questionId, @Bind("trueTemplateId") long trueTemplateId,
-               @Bind("falseTemplateId") long falseTemplateId);
+               @Bind("falseTemplateId") long falseTemplateId, long renderModeId);
```
Comment (MocanaAtBroad) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dao/JdbiBooleanQuestion.java]: We might need @Bind for renderModeId ?

Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dao/JdbiBooleanQuestion.java]: That we do! Thanks for the catch.

Comment (MocanaAtBroad) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dao/JdbiBooleanQuestion.java]: @ssettipalli Can you give this PR a quick look through? Do we have to anything additional for the export?

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/model/activity/types/BooleanRenderMode.java [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package org.broadinstitute.ddp.model.activity.types;
+
+public enum BooleanRenderMode {
+    RADIO_BUTTONS,
+    CHECKBOX;
```
Comment (Trimax) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/model/activity/types/BooleanRenderMode.java]: The semicolon is not required at the end of the enumeration

File: pepper-apis/dss-core/src/main/resources/org/broadinstitute/ddp/db/dao/JdbiQuestion.sql.stg [hunk 4]
```diff
@@ -97,6 +97,9 @@ select qt.question_type_code as question_type,
 
        bq.true_template_id,
        bq.false_template_id,
+       (select boolean_render_mode_code from boolean_render_mode
+         where boolean_render_mode_id = bq.boolean_render_mode_id
+       ) as boolean_render_mode,
```
Comment (Trimax) [inline pepper-apis/dss-core/src/main/resources/org/broadinstitute/ddp/db/dao/JdbiQuestion.sql.stg:102]: I'm not entirely sure, but if I'm not mistaken subselect queries are slower than joins.

```
explain select q.*,
       (select qt.question_type_code from question_type qt where qt.question_type_id = q.question_type_id)
from question q;
```
Снимок экрана 2022-05-10 в 16 04 18

vs 
```
explain select q.*, qt.question_type_code from question q
join question_type qt on qt.question_type_id = q.question_type_id;
```
Снимок экрана 2022-05-10 в 16 04 29

File: pepper-apis/dss-core/src/test/java/org/broadinstitute/ddp/db/dao/FormActivityDaoTest.java [hunk 5]
```diff
@@ -165,10 +165,11 @@ public void testInsertActivity_booleanQuestion() {
                 header,
                 footer,
                 rules,
+                true,
+                false,
                 yesOpt,
                 noOpt,
-                true,
-                false);
+                null); // Use the default render mode
```
Comment (Trimax) [inline pepper-apis/dss-core/src/test/java/org/broadinstitute/ddp/db/dao/FormActivityDaoTest.java]: What is the default render mode?

**Review by ssettipalli — DISMISSED** [review]
looks ok to me.

File: pepper-apis/dss-core/src/main/resources/org/broadinstitute/ddp/db/dao/JdbiQuestion.sql.stg (hunk 4 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/resources/org/broadinstitute/ddp/db/dao/JdbiQuestion.sql.stg:102]: My intention here was not to write the most optimal query, but to keep in style with the rest of the render_mode-style types used in the statement. If we want to refactor the whole thing, I'm game, but I don't think this is the place for it.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/model/activity/types/BooleanRenderMode.java (hunk 3 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/model/activity/types/BooleanRenderMode.java]: It is not, but it is within our current Checkstyle guidelines. If we want to flag this going forward, I suggest we enable the `UnnecessarySemicolonInEnumeration` check.

File: pepper-apis/dss-core/src/test/java/org/broadinstitute/ddp/db/dao/FormActivityDaoTest.java (hunk 5 above)
Comment (bskinner) [reply pepper-apis/dss-core/src/test/java/org/broadinstitute/ddp/db/dao/FormActivityDaoTest.java]: `RADIO_BUTTONS`. There will be a comment added to the `BooleanQuestionDef` constructor noting this behavior.

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/model/activity/types/BooleanRenderMode.java (hunk 3 above)
Comment (Trimax) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/model/activity/types/BooleanRenderMode.java]: I would vote for that!

**Review by Trimax — DISMISSED** [review]

**Review by Trimax — DISMISSED** [review]

**Review by Trimax — APPROVED** [review]

**Pull request merged**